File: Programmers/juyonLee/다리를지나는트럭.cpp
//[42583]
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;

    int curIdx = 0;
    int sum = 0;
    queue<int> q;

    while(1)
    {
        if(curIdx == truck_weights.size())
        {
            answer += bridge_length;
            break;
        }

        answer+=1;

        int curWeight = truck_weights[curIdx];

        if(q.size() == bridge_length)
        {
            sum -= q.front();
            q.pop();
        }

        if(sum + curWeight <= weight)
        {
            sum += curWeight;
            q.push(curWeight);
            curIdx+=1;
        }
        else
            q.push(0);
    }


    return answer;
}

//대기 순서대로 탐 + 다리 건너는 동안 무게 견디면 다음 자동차도 건너야 하므로 큐 이용.

File: Programmers/subin/다리를 지나는 트럭.py
def solution(bridge_length, weight, truck_weights):
    answer = 0
    track_idx = 0
    time = 0
    bridge_rest_w = weight
    bridge = []
    while True:
        if len(bridge) > 0 and time - bridge[len(bridge)-1][0] == bridge_length:
            bridge_rest_w += bridge[len(bridge)-1][1]
            bridge.pop()
        if track_idx < len(truck_weights) and bridge_rest_w >= truck_weights[track_idx]:
            bridge_rest_w  -= truck_weights[track_idx]
            bridge.insert(0,(time,truck_weights[track_idx]))
            track_idx += 1
        time += 1

        if len(bridge) == 0:
            break
    return time